src/RefractDataStructure.cc

```diff
@@ -1289,76 +1336,188 @@ namespace
+        return std::move(ref);
     }
 
-    std::unique_ptr<IElement> MsonElementToRefract(const NodeInfo<mson::Element>& mse,
+    std::unique_ptr<IElement> MsonTypeSectionToElements( //
+        const mson::Element::OneOfSection& section,
+        const snowcrash::SourceMap<mson::Element>* sourceMap,
         ConversionContext& context,
-        const mson::BaseTypeName defaultNestedType /* = mson::StringTypeName */)
+        const mson::BaseTypeName);
+
+    std::unique_ptr<IElement> MsonTypeSectionToElements( //
+        const mson::Element::OneOfSection& section,
+        const snowcrash::SourceMap<mson::Element>* sourceMap,
+        ConversionContext& context,
+        const mson::BaseTypeName)
     {
-        switch (mse.node->klass) {
-            case mson::Element::PropertyClass:
-                return MsonMemberToRefract<PropertyTrait>(
-                    MakeNodeInfo(mse.node->content.property, mse.sourceMap->property),
-                    context,
-                    GetType(mse.node->content.property.valueDefinition, context),
-                    defaultNestedType);
+        const NodeInfo<mson::OneOf>& oneOf = MakeNodeInfo(*section, sourceMap->oneOf());
 
-            case mson::Element::ValueClass:
-                return MsonMemberToRefract<ValueTrait>(MakeNodeInfo(mse.node->content.value, mse.sourceMap->value),
-                    context,
-                    GetType(mse.node->content.value.valueDefinition, context),
-                    defaultNestedType);
+        NodeInfoCollection<mson::OneOf> oneOfNodeInfo(oneOf);
+        auto select = oneOfNodeInfo.empty() ? make_empty<SelectElement>() : make_element<SelectElement>();
 
-            case mson::Element::MixinClass:
-                return MsonMixinToRefract(MakeNodeInfo(mse.node->content.mixin, mse.sourceMap->mixin));
+        for (const auto& oneOfInfo : oneOfNodeInfo) {
+            oneOfInfo.node->visit(MsonOneOfEntryToElementsLambda{ sourceMap, context, *select });
+        }
 
-            case mson::Element::OneOfClass:
-                return MsonOneofToRefract(MakeNodeInfo(mse.node->content.oneOf(), mse.sourceMap->oneOf()), context);
+        return std::move(select);
+    }
 
-            case mson::Element::GroupClass:
-                throw snowcrash::Error("unable to handle element group", snowcrash::ApplicationError);
+    std::unique_ptr<IElement> MsonTypeSectionToElements( //
+        const mson::Element::GroupSection&,
+        const snowcrash::SourceMap<mson::Element>*,
+        ConversionContext& context,
+        const mson::BaseTypeName)
+    {
+        // TODO avoid throwing exceptions...
+        throw snowcrash::Error("unable to handle element group", snowcrash::ApplicationError);
+    }
 
-            default:
-                throw snowcrash::Error("unknown type of mson element", snowcrash::ApplicationError);
+    struct MsonTypeSectionToElementsLambda {
```

Comment:
This can be a much shorter lambda in C++14. In C++11 we don't have a choice but introducing a structure with a template operator().